[simple_fsdp] Port fx pass to bucket reduce_scatters

IvanKobzarev opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #157780

Porting fx passes for reduce_scatters bucketing (similar to all_gather bucketing) for simple_fsdp and autoparallel testing.

cc @H-Huang @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @chenyang78 @kadeng @muchulee8 @amjames @chauhang @aakhundov

IvanKobzarev committed:
[simple_fsdp] Port fx pass to bucket reduce_scatters

[ghstack-poisoned]

wconstab commented on test/distributed/test_inductor_collectives.py:
```diff
@@ -1555,6 +1555,56 @@ def func(x, w, ag_0, ag_1, *, tag, ranks, group_size):
+        (FileCheck().check("reduce_scatter_tensor").run(code))
+        out = compiled(*inputs, **self.get_world_trs())
+        correct = func(*inputs, **self.get_world_trs())
+        assert same(out, correct), f"{out} va {correct}"
```
shouldn't this test also check that there is only one RS in the graph and confirm the reordering pass did the reordering?

IvanKobzarev committed:
Update on "[simple_fsdp] Port fx pass to bucket reduce_scatters"


Porting fx passes for reduce_scatters bucketing (similar to all_gather bucketing) for simple_fsdp and autoparallel testing.

cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

wconstab commented on test/distributed/test_inductor_collectives.py:
```diff
@@ -1676,21 +1683,49 @@ def _reorder_communication_preserving_peak_memory(
-        (FileCheck().check("all_gather_into_tensor_out").run(code))
+        (
+            FileCheck()
+            .check_count(
```
just out of curiosity, any reason you used 3 separate FileCheck instances instead of just doing it all in one?

IvanKobzarev replied:
Oh, thanks, it can be done as a one check :)

IvanKobzarev committed:
Update on "[simple_fsdp] Port fx pass to bucket reduce_scatters"


Porting fx passes for reduce_scatters bucketing (similar to all_gather bucketing) for simple_fsdp and autoparallel testing.

cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]
